src/toxcore/dht/server/mod.rs

```diff
@@ -74,6 +75,25 @@ struct ServerBootstrapInfo {
     motd_cb: Arc<Fn(&Server) -> Vec<u8> + Send + Sync>,
 }
 
+/// Struct that contains PrecomputedKey of a peer and last time it is accessed.
+#[derive(Clone)]
+pub struct PrecomputedKeyAndTime {
+    /// PrecomputedKey of a peer.
+    precomputed_key: PrecomputedKey,
+    /// last time of accessing precomputed_key.
+    pub last_time: Instant,
```

Comment:
Why do you keep `last_time`? What's that for?

Reply:
It is for removing precomputed keys.
If last_time span is greater than KILL_NODE_TIMEOUT then it is removed.

Reply:
You don't need to duplicate it. Once you remove the node you can remove the key by node' PK

Reply:
https://github.com/tox-rs/tox/blob/9309968c7ae68dccfc9b22a1687beea7989e67c2/src/toxcore/dht/server/mod.rs#L763-L765

When we handle ping_request we create and save precomputed key.
But it is possible we don't add peer into `nodes_to_ping` by `ping_add` if the peer can not be added to close_node.
This means that we just responds with ping_response, no ping_id was created, no node is added to close_nodes.
So, we need to maintain separated `last_time` field. Right?

Reply:
Then we can maintain precomputed key only for addable node to close_nodes.

Reply:
I'd suggest to use https://github.com/jeromefroe/lru-rs
We can't evict precomputed keys using time policy only - it can lead to OOM if someone decides to dos us with different keys. So we have to set limit anyway. But if we have limit and eviction strategy based on size we don't need to store time anymore.

Reply:
Another way we can handle it - store precomputed keys inside `DhtNode`. But this will work only for nodes that we have in our `Ktree`.

Reply:
I prefer `LRU' @kpp ?

Reply:
Or perhaps we can have global cache using https://github.com/jaemk/cached

Reply:
It is needed to test which case is more efficient, one case is not using precomputed keys, the other case is using `LRU` or storing precomputed key inside `DhtNode`